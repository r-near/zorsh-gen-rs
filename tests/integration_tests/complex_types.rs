use anyhow::Result;

#[test]
fn test_nested_structs() -> Result<()> {
    let input = r#"
        #[derive(BorshSerialize)]
        struct Address {
            street: String,
            city: String,
            country: String,
        }

        #[derive(BorshSerialize)]
        struct User {
            name: String,
            address: Address,
        }
    "#;

    let output = zorsh_gen_rs::convert_str(input)?;
    insta::assert_snapshot!(output);
    Ok(())
}

#[test]
fn test_collections() -> Result<()> {
    let input = r#"
        #[derive(BorshSerialize)]
        struct Item {
            id: u32,
            name: String,
        }

        #[derive(BorshSerialize)]
        struct Inventory {
            items: Vec<Item>,
            counts: HashMap<String, u32>,
            fixed_data: [u8; 32],
        }
    "#;

    let output = zorsh_gen_rs::convert_str(input)?;
    insta::assert_snapshot!(output);
    Ok(())
}

#[test]
fn test_complex_enums() -> Result<()> {
    let input = r#"
        #[derive(BorshSerialize)]
        enum Event {
            Created { timestamp: u64 },
            Updated(String),
            Deleted,
        }

        #[derive(BorshSerialize)]
        struct EventLog {
            events: Vec<Event>,
        }
    "#;

    let output = zorsh_gen_rs::convert_str(input)?;
    insta::assert_snapshot!(output);
    Ok(())
}

#[test]
fn test_optional_and_nested_types() -> Result<()> {
    let input = r#"
        #[derive(BorshSerialize)]
        struct Metadata {
            tags: Vec<String>,
            extra: Option<HashMap<String, Vec<u32>>>,
        }

        #[derive(BorshSerialize)]
        struct Document {
            id: String,
            metadata: Option<Metadata>,
        }
    "#;

    let output = zorsh_gen_rs::convert_str(input)?;
    insta::assert_snapshot!(output);
    Ok(())
}
