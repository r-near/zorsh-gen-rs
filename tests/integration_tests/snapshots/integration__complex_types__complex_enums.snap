---
source: tests/integration_tests/complex_types.rs
expression: output
snapshot_kind: text
---
import { b } from '@zorsh/zorsh';

export const EventSchema = b.enum({
    Created: b.struct({
        timestamp: b.u64()
    }),
    Updated: b.string(),
    Deleted: b.unit()
});

export const EventLogSchema = b.struct({
    events: b.vec(EventSchema)
});

export {
    EventSchema,
    EventLogSchema
};
